paddle/fluid/operators/setitem_op.cc

```diff
@@ -0,0 +1,104 @@
+
+#include <string>
+
+namespace paddle {
+namespace operators {
+
+class SetitemOp : public framework::OperatorWithKernel {
+ public:
+  SetitemOp(const std::string &type, const framework::VariableNameMap &inputs,
+            const framework::VariableNameMap &outputs,
+            const framework::AttributeMap &attrs)
+      : OperatorWithKernel(type, inputs, outputs, attrs) {}
+
+  void InferShape(framework::InferShapeContext *ctx) const override {
+    OP_INOUT_CHECK(ctx->HasInput("Input"), "Input", "Input", "Setitem");
+    OP_INOUT_CHECK(ctx->HasOutput("Out"), "Output", "Out", "Setitem");
+    auto in_dims = ctx->GetInputDim("Input");
+    PADDLE_ENFORCE_LT(
+        in_dims.size(), 7,
+        platform::errors::InvalidArgument(
+            "The rank of input should be less than 7, but received %d.",
+            in_dims.size()));
+  }
+
+ protected:
+  framework::OpKernelType GetExpectedKernelType(
+      const framework::ExecutionContext &ctx) const override {
+    return framework::OpKernelType(
+        framework::proto::VarType::Type(ctx.Attr<int>("dtype")),
+        ctx.GetPlace());
+  }
+};
+
+class SetitemOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  void Make() override {
+    AddInput("Input", "(Tensor) Input tensor of setitem operator.");
+    AddInput("ValueTensor", "(Tensor) Value tensor of setitem operator.")
+        .AsDispensable();
+    AddOutput("Out",
+              "(Tensor) Output tensor of setitem operator. The output is the "
+              "same Tensor as input");
+
+    AddAttr<int>("dtype", "data type of input.")
+        .InEnum(
+            {framework::proto::VarType::BOOL, framework::proto::VarType::INT32,
+             framework::proto::VarType::INT64, framework::proto::VarType::FP32,
+             framework::proto::VarType::FP64})
+        .SetDefault(framework::proto::VarType::FP32);
+    AddAttr<std::vector<int>>(
+        "axes", "(list<int>) Axes that `starts` and `ends` apply to.");
+    AddAttr<std::vector<int>>(
+        "starts",
+        "(list<int>) Starting indices of corresponding axis in `axes`");
+    AddAttr<std::vector<int>>(
+        "ends", "(list<int>) Ending indices of corresponding axis in `axes`.");
+
+    AddAttr<std::vector<int>>("bool_values", "store the bool values")
+        .SetDefault({});
+    AddAttr<std::vector<float>>("fp32_values", "store the float32 values")
+        .SetDefault({});
+    AddAttr<std::vector<int>>("int32_values", "store the int32 values")
+        .SetDefault({});
+    AddAttr<std::vector<int64_t>>("int64_values", "store the int64 values")
+        .SetDefault({});
+
+    AddAttr<std::vector<int>>("shape", "(vector<int>) Shape of values.")
+        .SetDefault({});
+    AddComment(R"DOC(Setitem operator.
+Assignment to a Tensor in static mode.
+)DOC");
+  }
+};
+}  // namespace operators
+}  // namespace paddle
+
+namespace ops = paddle::operators;
+
+REGISTER_OPERATOR(
+    setitem, ops::SetitemOp, ops::SetitemOpMaker,
```

Comment:
Should we name the operator "set_item" to match C++ style instead of Python style?

Reply:
Yes, thanks. Now replace "set_item" with "set_value".